### DISCUSSION TITLE
Nothing

### DISCUSSION THREAD

**Discussion by steven-ld** [Keyfactor/ejbca-ce#725] [Q&A]
nothing

**Comment by primetomas** [comment]
Which API do you want to use? EJBCA includes clientToolBox that is WebService CLI, with plenty of sample code.
https://github.com/Keyfactor/ejbca-ce/tree/main/modules/ejbca-ws-cli

For the REST API we have an open source command line client as well, all code is here: https://github.com/Keyfactor/ejbca-easy-rest-client

**Reply by steven-ld** [reply]
![image](https://github.com/user-attachments/assets/a736b06f-808c-4f58-9a38-cdeb44a66fe2)
There are some problems with the compilation. Is there a compiled jar?

**Reply by primetomas** [reply]
That was a maven issue. I fixed it so please update and try again.

**Reply by steven-ld** [reply]
Thank you for your help. It seems that can communicate with EJBCA now. I watched this youtube video https://www.youtube.com/watch? v=wMD1GgSF-JE, I first cloned the profiles of enduser, and then created a Profile of End Entity of ClientAuth.

```shell
(base) ➜  ejbca-easy-rest-client git:(main) ✗ java -jar target/erce-0.3.0-SNAPSHOT.jar enroll genkeys \
--authkeystore /Users/ga666666/Desktop/SuperAdmin.p12 \
--authkeystorepass 54lE9avyVVlut72lUunWQV8m \
--endentityprofile "ClientAuth" \
--certificateprofile "ClientAuth" \
--ca ManagementCA \
--subjectaltname "dnsName=test-erces-01.test" \
--hostname localhost \
--destination ./certs \
--subjectdn "C=SE, O=Keyfactor, CN=test-erces-01.test" \
--username test-erces-01.test -p \
--keyalg EC --keyspec P-256 --verbose

SETTING: --authkeystore as /Users/ga666666/Desktop/SuperAdmin.p12
SETTING: --authkeystorepass as 54lE9avyVVlut72lUunWQV8m
SETTING: --endentityprofile as ClientAuth
SETTING: --certificateprofile as ClientAuth
SETTING: --ca as ManagementCA
SETTING: --subjectaltname as dnsName=test-erces-01.test
SETTING: --hostname as localhost
SETTING: --destination as ./certs
SETTING: --subjectdn as C=SE, O=Keyfactor, CN=test-erces-01.test
SETTING: --username as test-erces-01.test
Enter password for parameter (-p):

SETTING: --keyalg as EC
SETTING: --keyspec as P-256
Generating named curve ECDSA key pair: P-256
Generated EC keys with length 256
Using default values for DN components
Properties file /profilemappings.properties was not found.
Properties file /profilemappings_enterprise.properties was not found.
Returning ecAlgs: [SHA256withECDSA, SHA1withECDSA, SHA224withECDSA, SHA384withECDSA, SHA512withECDSA, SHA3-256withECDSA, SHA3-384withECDSA, SHA3-512withECDSA]
SigAlg: SHA256withECDSA
Return code was: 400: {"error_code":400,"error_message":"Wrong number of ORGANIZATION fields in Subject DN."}
Exception in thread "main" java.lang.IllegalArgumentException: Parameter certificate cannot be null.
	at com.keyfactor.ejbca.client.enroll.EnrollGenKeysCommand.createP12(EnrollGenKeysCommand.java:188)
	at com.keyfactor.ejbca.client.enroll.EnrollGenKeysCommand.execute(EnrollGenKeysCommand.java:148)
	at com.keyfactor.ejbca.client.ErceCommandBase.execute(ErceCommandBase.java:113)
	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:309)
	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:320)
	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary.findAndExecuteCommandFromParameters(CommandLibrary.java:88)
	at com.keyfactor.ejbca.client.ErceClient.main(ErceClient.java:32)
(base) ➜  ejbca-easy-rest-client git:(main) ✗
```

What other modifications do I need to make? Thank you for your help.

**Reply by primetomas** [reply]
Great. "Wrong number of ORGANIZATION fields in Subject DN" is a policy validation message from the server. It means that your end entity profile (ClientAuth) probably does not include the O field (or have more than one O field required), and you have an O in the requested subjectDN (O=Keyfactor).

**Reply by primetomas** [reply]
Select O in the subject DN attributes drop down and click add, then save. Do the same for all other attributes you want to include.

**Reply by dobicinaitis** [reply]
If you want to interact with EJBCA from a Spring Boot application, the [ejbca-java-client-sdk](https://github.com/Keyfactor/ejbca-java-client-sdk) might suit your needs better.

**Reply by primetomas** [reply]
I'd assume that clientcertificate should be the, in this case superadmin, certificate in PEM format. And "key" should be the same "PRIVATE KEY" exported from the P12 file as you mention. Maybe it has to be without the password, "--nodes" option to "openssl pkcs12".

**Reply by primetomas** [reply]
Same as #726
